### ISSUE TITLE
nix-env -u fails with assertion in attr-set.hh

### ISSUE THREAD

**Issue by dsxmachina** [NixOS/nix#4709]
**Describe the bug**

Updating user packages fails with cryptic error message:
```
> nix-env -u
upgrading 'newsboat-2.20.1' to 'newsboat-2.23'
upgrading 'dunst-1.5.0' to 'dunst-1.6.1'
upgrading 'lsof-4.93.2' to 'lsof-4.94.0'
upgrading 'rofi-calc-1.8' to 'rofi-calc-2.0.0'
upgrading 'rofi-1.5.4' to 'rofi-1.6.1'
upgrading 'mpv-0.32.0' to 'mpv-0.33.1'
upgrading 'picom-8' to 'picom-8.2'
...
nix-env: src/libexpr/attr-set.hh:54: void nix::Bindings::push_back(const nix::Attr&): Assertion `size_ < capacity_' failed.
```
The list of packages is way bigger, but it seems that the bug is unrelated to the specific packages

**Steps To Reproduce**

Type in "nix-env -u" and watch.

**Expected behavior**

Update of my user-packages.

**`nix-env --version` output**
nix-env (Nix) 2.3.10

**Additional context**
Nixos 20.09, User is subscribed to stable (20.09) and unstable channel for different packages.

**Comment by remexre** [comment]
I reproduce this for the neovim package on aarch64 qemu nixos, build 142244707

**Comment by rkoe** [comment]
Reproduced with a fresh single-user-installation of Nix, with chromium:

```
$ nix-env -iA nixpkgs.chromium
installing 'chromium-90.0.4430.93'
nix-env: src/libexpr/attr-set.hh:54: void nix::Bindings::push_back(const nix::Attr&): Assertion `size_ < capacity_' failed.
Aborted
```

**Comment by efx** [comment]
Notes for the curious.
This assertion fails:
https://github.com/NixOS/nix/blob/fe3a10a9b2c61ba19b24da4c93c4abc9e7583f3d/src/libexpr/attr-set.hh#L56

You can install prior versions of nix with:
```sh
 curl -L https://releases.nixos.org/nix/nix-2.3.7/install | sh
```
I've tried back to 2.3.7 and receive the same error when I run `nix-env -iA nixpkgs.chromium`.

**Comment by efx** [comment]
I wonder if the cause of the bug is the length of chromium's version. e.g. `chromium-90.0.4430.93`. Running `nix-env -iA nixpkgs.firefox` works fine.

**Comment by efx** [comment]
> I wonder if the cause of the bug is the length of the chromium package version. e.g. chromium-90.0.4430.93. Running nix-env -iA nixpkgs.firefox works fine.

That is not the case.

I can install a version of chromium from nixos-20.09 with the latest version of nix (nix-2.3.10).
Here were my steps:
1. reinstall nix `curl -L https://nixos.org/nix/install | sh`
1. `nix-channel --add https://nixos.org/channels/nixos-20.09 nixos`
1. `nix-channel --update`
1. `nix-env -iA nixos.chromium`

**Comment by pinkisemils** [comment]
I can reproduce with `nix-env -if my-packages.nix`.

**Comment by vcunat** [comment]
Fixed by #4411 (stable Nix 2.3.11).
